kola/tests/update/update.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2018 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package update
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-omaha/omaha"
+
+	"github.com/coreos/mantle/kola"
+	"github.com/coreos/mantle/kola/cluster"
+	"github.com/coreos/mantle/kola/register"
+	tutil "github.com/coreos/mantle/kola/tests/util"
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/platform/local"
+	"github.com/coreos/mantle/platform/machine/qemu"
+	"github.com/coreos/mantle/util"
+)
+
+func init() {
+	register.Register(&register.Test{
+		Name:        "coreos.update.payload",
+		Run:         payload,
+		ClusterSize: 1,
+		NativeFuncs: map[string]func() error{
+			"Omaha": Serve,
+		},
+	})
+}
+
+func Serve() error {
+	omahaserver, err := omaha.NewTrivialServer(":34567")
+	if err != nil {
+		return fmt.Errorf("creating trivial omaha server: %v\n", err)
+	}
+
+	omahawrapper := local.OmahaWrapper{TrivialServer: omahaserver}
+
+	if err = omahawrapper.AddPackage("/updates/update.gz", "update.gz"); err != nil {
+		return fmt.Errorf("bad payload: %v", err)
+	}
+
+	return omahawrapper.Serve()
+}
+
+func payload(c cluster.TestCluster) {
+	addr := ConfigureOmahaServer(c, c.Machines()[0])
+
+	// create the actual test machine, the machine
+	// that is created by the test registration is
+	// used to host the omaha server
+	m, err := c.NewMachine(nil)
+	if err != nil {
+		c.Fatalf("creating test machine: %v", err)
+	}
+
+	ConfigureMachineForUpdate(c, m, addr)
+
+	tutil.AssertBootedUsr(c, m, "USR-A")
+
+	UpdateMachine(c, m)
+
+	tutil.AssertBootedUsr(c, m, "USR-B")
+
+	tutil.InvalidateUsrPartition(c, m, "USR-A")
+
+	UpdateMachine(c, m)
+
+	tutil.AssertBootedUsr(c, m, "USR-A")
+}
+
+func ConfigureOmahaServer(c cluster.TestCluster, srv platform.Machine) string {
+	if kola.UpdatePayloadFile == "" {
+		c.Skip("no update payload provided")
+	}
+
+	if qc, ok := c.Cluster.(*qemu.Cluster); ok {
+		// An Omaha Server is built for all qemu clusters via the LocalCluster;
+		// Destroy it before setting up the omaha server on the machine created
+		// for the test.
+		qc.OmahaServer.Destroy()
```

Comment:
Why is this necessary?  Also, is `OmahaServer` safe against multiple destructions?

Reply:
I'll test without destroying this, I left the disabling code because previously it was overriding the `OmahaServer` I was setting up but that shouldn't be an issue anymore.

Reply:
You were correct that this is no longer necessary. Now that the `OmahaServer` is being hosted on the original machine created by `kola` it is on a different IP than the host (`10.0.0.2` for the `kola` machine vs `10.0.0.1` for the host).

Pushed an update dropping this.